### PR TITLE
KCL-11106 - Update selection before handling input in case the select…

### PR DESCRIPTION
**Pull request by hejtmii**
See details in task

### CODE REVIEW

File: src/component/handlers/edit/DraftEditorEditHandler.js [hunk 1]
```diff
@@ -31,11 +32,55 @@ const onSelect = require('editOnSelect');
 const isChrome = UserAgent.isBrowser('Chrome');
 const isFirefox = UserAgent.isBrowser('Firefox');
 
-const selectionHandler: (e: DraftEditor) => void =
-  isChrome || isFirefox ? onSelect : e => {};
+// In certain cases, contenteditable on chrome does not fire the onSelect
+// event, causing problems with cursor positioning. Therefore, the selection
```
Comment (vilemj-Viclick) [inline src/component/handlers/edit/DraftEditorEditHandler.js]:
```
// event, causing problems with cursor positioning. Therefore, the selection
                                  ↑ Did you mean "caret"?
```

File: src/component/handlers/edit/DraftEditorEditHandler.js [hunk 2]
```diff
@@ -31,11 +32,55 @@ const onSelect = require('editOnSelect');
...
+// state update handler is added to more events to ensure that the selection
+// state is always synced with the actual cursor positions.
```
Comment (vilemj-Viclick) [inline src/component/handlers/edit/DraftEditorEditHandler.js]:
```
// state is always synced with the actual cursor positions.
                                      ↑ current?
```

File: src/component/handlers/edit/DraftEditorEditHandler.js [hunk 3]
```diff
@@ -31,11 +32,55 @@ const onSelect = require('editOnSelect');
...
+const explicitlyHandleCaretPositioning = isChrome || isFirefox;
+
+const keysDown: Set<number> = new Set();
+
+function updateTrackedKeyStatus(
+  e: SyntheticKeyboardEvent<HTMLElement>,
+  isDown: boolean,
+) {
+  const keyCode = e.which;
```
Comment (vilemj-Viclick) [inline src/component/handlers/edit/DraftEditorEditHandler.js:34]: Aside from "e.which" being deprecated, this also works for detecting numpad keys when num-lock is off.
I could not think of any other key which would affect selection aside from "Meta". The key with the Windows logo on it. But that one basically takes focus away from the window entirely, so it's not that much of a problem, is it?

File: src/component/handlers/edit/DraftEditorEditHandler.js [hunk 4]
```diff
@@ -31,11 +32,55 @@ const onSelect = require('editOnSelect');
...
+// event, causing problems with cursor positioning. Therefore, the selection
+// state update handler is added to more events to ensure that the selection
+// state is always synced with the actual cursor positions.
+const explicitlyHandleCaretPositioning = isChrome || isFirefox;
+
+const keysDown: Set<number> = new Set();
+
+function updateTrackedKeyStatus(
+  e: SyntheticKeyboardEvent<HTMLElement>,
+  isDown: boolean,
```
Comment (vilemj-Viclick) [inline src/component/handlers/edit/DraftEditorEditHandler.js]: I don't like this boolean toggle, honestly.

Why not just have
```
function setTrackedKeyStatus(
  e: SyntheticKeyboardEvent<HTMLElement>,
) {
  // Same implementation except the "if"
};
```

and then reset the key on line 112 simply by
```
keysDown.delete(e.which);
```

Comment (vilemj-Viclick) [reply src/component/handlers/edit/DraftEditorEditHandler.js]: Also... The variable called `keyCode` when you're specifically ignoring `e.keyCode` had me looking at the code in a wrong way for a while. I thought you were using `e.keycode` to tell apart keys. But they are practically equivalent. Both are deprecated. :D

**Review by vilemj-Viclick — APPROVED** [review]
It's basically fine.
Just some nitpicks.

File: src/component/handlers/edit/DraftEditorEditHandler.js (hunk 3 above)
Comment (hejtmii) [reply src/component/handlers/edit/DraftEditorEditHandler.js:34]: I guess not. e.which is also used in keyboard shortcuts handling in other parts of the code, so if it stops working, we are in much larger trouble than just this :-)

File: src/component/handlers/edit/DraftEditorEditHandler.js (hunk 4 above)
Comment (hejtmii) [reply src/component/handlers/edit/DraftEditorEditHandler.js]: True, I wanted to keep implementation details of how the keys are tracked in the same place. I got rid of the IF and made two methods instead

Regarding the `keyCode`, I literally copied the code from `editOnKeyDown.js` to keep things consistent with existing code that works in a similar way and with the same Keys constants

File: src/component/handlers/edit/DraftEditorEditHandler.js (hunk 2 above)
Comment (hejtmii) [reply src/component/handlers/edit/DraftEditorEditHandler.js]: updated

File: src/component/handlers/edit/DraftEditorEditHandler.js (hunk 1 above)
Comment (hejtmii) [reply src/component/handlers/edit/DraftEditorEditHandler.js]: This is comment from original DraftJS code, just move higher, I didn't want to touch it on purpose. Keeping it

**Review by vilemj-Viclick — APPROVED** [review]
Now it's even more fine. 😄 
:shipit:

**Review by vilemj-Viclick — CHANGES_REQUESTED** [review]
image
Please remove the browser specification here as we handle these events in all of them and we know for a fact it also happens in other browsers.

**Review by vilemj-Viclick — APPROVED** [review]

**Pull request merged**